File: nullobjectpattern/nullobjectpattern.py
from abc import ABC, abstractmethod

class Vehicle(ABC):
    @abstractmethod
    def getTankCapacity(self):
        pass

    @abstractmethod
    def getSeatingCapacity(self):
        pass

class VehicleFactory:

    @staticmethod
    def getVehicleObject(typeOfVehicle):
        if typeOfVehicle == 'Car':
            return Car()
        return NullVehicle()

class NullVehicle(Vehicle):
    def getTankCapacity(self):
        return 0

    def getSeatingCapacity(self):
        return 0

class Car(Vehicle):
    def getTankCapacity(self):
        return 40

    def getSeatingCapacity(self):
        return 5

if __name__ == '__main__':
    vehicle = VehicleFactory.getVehicleObject('Bike')
    # vehicle = VehicleFactory.getVehicleObject('Car')
    print("Seating Capacity: ", vehicle.getSeatingCapacity())
    print("Fuel Tank Capacity: ", vehicle.getTankCapacity())

File: nullobjectpattern/nullobjectpattern.js
// Abstract base class (simulated)
class Vehicle {
  getTankCapacity() {
    throw new Error("Method 'getTankCapacity()' must be implemented");
  }

  getSeatingCapacity() {
    throw new Error("Method 'getSeatingCapacity()' must be implemented");
  }
}

class VehicleFactory {
  static getVehicleObject(typeOfVehicle) {
    if (typeOfVehicle === "Car") {
      return new Car();
    }
    return new NullVehicle();
  }
}

class NullVehicle extends Vehicle {
  getTankCapacity() {
    return 0;
  }

  getSeatingCapacity() {
    return 0;
  }
}

class Car extends Vehicle {
  getTankCapacity() {
    return 40;
  }

  getSeatingCapacity() {
    return 5;
  }
}

// ---- Client code ----
const vehicle = VehicleFactory.getVehicleObject("Bike");
// const vehicle = VehicleFactory.getVehicleObject("Car");

console.log("Seating Capacity:", vehicle.getSeatingCapacity());
console.log("Fuel Tank Capacity:", vehicle.getTankCapacity());
